From 29b0bab5f28edf2741c763f5f1ecacf63c66046d Mon Sep 17 00:00:00 2001
From: Ondrej Kozina <okozina@redhat.com>
Date: Mon, 8 Sep 2014 19:03:56 -0400
Subject: [crypto] algif: avoid excessive use of socket buffer in skcipher

Message-id: <540DFD9C.6010300@redhat.com>
Patchwork-id: 92481
O-Subject: Subject: [RHEL7.1 PATCH] crypto: algif - avoid excessive use of socket buffer in skcipher
Bugzilla: 1136075
RH-Acked-by: Jiri Benc <jbenc@redhat.com>
RH-Acked-by: Herbert Xu <herbert.xu@redhat.com>

BZ: 1136075

This patch is essential for cryptsetup rebase planned/approved for 7.1.
Also it'll fix the already existing issue preventing actual cryptsetup
version opening TrueCrypt containers on ppc64.

Fix propsed for upstream:
http://www.mail-archive.com/linux-crypto@vger.kernel.org/msg11787.html

Herbert already staged the patch for 3.18 inclusion.

Upstream cryptodev-2.6.git commit e2cffb5f493a8b431dc87124388ea59b79f0bccb
Author: Ondrej Kozina <okozina@redhat.com>
Date:   Mon Aug 25 11:49:54 2014 +0200

     crypto: algif - avoid excessive use of socket buffer in skcipher

     On archs with PAGE_SIZE >= 64 KiB the function skcipher_alloc_sgl()
     fails with -ENOMEM no matter what user space actually requested.
     This is caused by the fact sock_kmalloc call inside the function tried
     to allocate more memory than allowed by the default kernel socket buffer
     size (kernel param net.core.optmem_max).

     Signed-off-by: Ondrej Kozina <okozina@redhat.com>
     Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/crypto/algif_skcipher.c b/crypto/algif_skcipher.c
index a19c027..83187f4 100644
--- a/crypto/algif_skcipher.c
+++ b/crypto/algif_skcipher.c
@@ -49,7 +49,7 @@ struct skcipher_ctx {
  struct ablkcipher_request req;
 };
 
-#define MAX_SGL_ENTS ((PAGE_SIZE - sizeof(struct skcipher_sg_list)) / \
+#define MAX_SGL_ENTS ((4096 - sizeof(struct skcipher_sg_list)) / \
         sizeof(struct scatterlist) - 1)
 
 static inline int skcipher_sndbuf(struct sock *sk)
-- 
1.7.1